### PR TITLE
Add STL compliant field iteration to MIMEHdr.

### PR DESCRIPTION
**Pull request by SolidWallOfCode**
Changes for issue #7472. This removes the old "iter" methods and replaces them with an STL compliant iterator. Uses of the old methods are updated to use the new iterator.

### CODE REVIEW

File: proxy/http/HttpTransact.cc [hunk 1]
```diff
@@ -5005,31 +5002,20 @@ HttpTransact::merge_response_header_with_cached_header(HTTPHdr *cached_header, H
...
-        while (dname) {
-          cached_header->field_delete(dname, dlen);
-          dfield = response_header->iter_get_next(&fiter2);
-          if (dfield) {
-            dname = dfield->name_get(&dlen);
-          } else {
-            dname = nullptr;
-          }
+        for (auto spot2 = spot; spot2 != limit; ++spot2) {
+          cached_header->field_delete(&*spot2, true);
```
Comment (SolidWallOfCode) [inline proxy/http/HttpTransact.cc:5011]: Not completely sure about this change - is it faster to delete via the dup chain, or continue on the iteration over all of the remaining headers?

File: proxy/hdrs/MIME.h [hunk 2]
```diff
@@ -987,6 +987,43 @@ struct MIMEFieldIter {
...
+   */
+  class iterator
+  {
+    using self_type = iterator; ///< Self reference types.
+
+  public:
+    iterator() = default;
+
+    // STL iterator compliance types.
+    using difference_type   = void;
```
Comment (dragon512) [inline proxy/hdrs/MIME.h:1001]: Should this not be size_t ?

Comment (SolidWallOfCode) [reply proxy/hdrs/MIME.h:1001]: No, because it's not possible to compute a difference.

File: proxy/http/HttpTransact.cc (hunk 1 above)
Comment (bigcook12345) [reply proxy/http/HttpTransact.cc:5013]: Headers were input for the commits that were added. Would it have passed in the system and or test otherwise?

**Review by bigcook12345 — APPROVED** [review]
Headers due to each commit. I'm not sure if it's "Faster" per say..?  But surely passing in the system. To keep the commits uniform...

File: proxy/hdrs/MIME.h (hunk 2 above)
Comment (bigcook12345) [reply proxy/hdrs/MIME.h:1001]: So, is there a way to run it with out deleting what's been inserted..? With out pulling it apart?

**Review by bigcook12345 — COMMENTED** [review]
Just a thought...

**Comment by SolidWallOfCode** [comment]
[ci approve ubuntu]

**Comment by bneradt** [comment]
[approve ci clang-analyzer]

**Review by maskit — APPROVED** [review]
Looks good to me. I ran test_HPACK a couple of time with and without this change and there was no execution time difference.

**Pull request merged**